.github/workflows/maven.yml

```diff
@@ -0,0 +1,116 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: Java CI
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+        java: [8, 11, 14]
+      fail-fast: false
+
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set up cache for ~./m2/repository
+        uses: actions/cache@v1
+        with:
+          path: ~/.m2/repository
+          key: maven-${{ matrix.os }}-java${{ matrix.java }}-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            maven-${{ matrix.os }}-java${{ matrix.java }}-
```

Comment:
You might want to consider using a version number in the cache key. If - for any reason - the current caches become unsuitable, you can bump the version number and thus manually invalidate it.

Reply:
Hmm, interesting. How it currently works is that it will try to find a cache exactly matching the key: `maven-${{ matrix.os }}-java${{ matrix.java }}-${{ hashFiles('**/pom.xml') }}`. If that fails it will try to find any of the `restore-keys` caches (which are prefixes of caches). So it is designed to always try to find any maven repository cache.

There is no way to manually clear the cache as far as I know, so if we really want this to work, we should remove the `restore-keys` list and always hit on exactly the cache key. If we would need to clear the cache, we can make a simple change in a pom file which changes the hash in the caching key. It's a bit cumbersome in my opinion.

Alternatively we could also keep it like this and make a change once needed. We can always remove the caching job in this file, when needed. Also I think that the GitHub Actions are only advisory (they won't block PRs), so we could always fallback to local runs.

Reply:
Hmm, I see, apparently the caching mechanism works differently than it does in CircleCI (where I have often seen the practice of prefixing a cache with a version identifier. Let's keep it as it is.